src/ocean/math/WideInt.d

```diff
@@ -0,0 +1,474 @@
+
+    public void opAddAssign ( uint rhs )
+    {
+        if (add(this.payload[0], rhs))
+            enforce(.wideint_exception, checkAndInc(1));
+    }
+
+    unittest
+    {
+        WideInt num = 0;
+        for (auto i = 0; i < 1000; ++i)
+            num += (1 << 31);
+        test!("==")(num.payload[0], 0);
+        test!("==")(num.payload[1], 500);
+    }
+
+    /***************************************************************************
+
+        Mutates current number in-place, dividing it by `rhs` and calculating
+        remainder.
+
+        Params:
+            rhs = number to divide by
+
+        Returns:
+            remainder of division
+
+    ***************************************************************************/
+
+    public uint divideBy ( uint rhs )
+    {
+        ulong remainder = 0;
+
+        for (ptrdiff_t idx = this.payload.length - 1; idx >= 0; --idx)
+        {
+            remainder = (remainder << 32) + this.payload[idx];
+            ulong result = remainder / rhs;
+            remainder -= rhs * result;
+            assert(result <= uint.max);
+            this.payload[idx] = cast(uint) result;
+        }
+
+        assert(remainder <= uint.max);
+        return cast(uint) remainder;
+    }
+
+    unittest
+    {
+        WideInt num = 56 * 47;
+        test!("==")(0, num.divideBy(47));
+        test!("==")(num.payload[0], 56);
+
+        num.payload[0] = 1337;
+        num.payload[1] = 100;
+        test!("==")(1337, num.divideBy(1 << 31));
+        test!("==")(num.payload[0], 200);
+    }
+
+    /***************************************************************************
+
+        Mutates this number in place multiplying it by a regular integer
+
+        Params:
+            rhs = value to multiply by
+
+        Throws:
+            WideIntRangeException if resulting value can't fit in current
+            WideInt value range
+
+    ***************************************************************************/
+
+    public void multiplyBy ( uint rhs )
+    {
+        ulong overflow = 0;
+
+        for (size_t i = 0; i < this.payload.length; ++i)
+        {
+            overflow += cast(ulong)(this.payload[i]) * rhs;
+            this.payload[i] = cast(uint) (overflow & uint.max);
+            overflow >>= 32;
+        }
+
+        enforce(.wideint_exception, overflow == 0);
+    }
+
+    unittest
+    {
+        WideInt num = 1_000_000_000;
+        num.multiplyBy(1_000_000_000);
+        test!("==")(num.toString(), "1000000000000000000");
+    }
+
+    /***************************************************************************
+
+        Returns:
+            double precision floating point representation of this number
+
+    ***************************************************************************/
+
+    public double toDouble ( )
+    {
+        // TODO: for widenumbers with value > ulong.max need to calculate
+        // mantiss and exponent manually by counting bitshift
```

Comment:
If you know that the payload length is always small, you can just use the brute force approach.

```d
double result = 0;
foreach_reverse (word, this.payload )
{
   result *= 2^^32; // this is an exactly representable number, it's an error-free calculation
   result += word;
}
```

If that's too slow you can speed it up by scanning for the first non-zero payload word. There are only 63 bits of mantissa in `real` so additions after the first three have no effect and can be skipped. (We need 3 words, not 2, because we might only get 1 bit from the first payload word).
If `p` is the index of the first non-zero word, then

```d
const double MAXWORD = 2.0^^32; // so we get floating-point multiplies. Could also use real instead`
return (((this.payload[p]*MAXWORD +(this.payload[p-1]))*MAXWORD)+this.payload[p-2])* ldexp(1.0, (p-2)*32);
```

Of course you need to worry about array bounds errors if p is less than 2. But you can set `p = min(p, 2)` because the formula is still valid even if the high payload values are zero.